### ISSUE TITLE
Cleaning up effects from removed widget

### ISSUE THREAD

**Issue by justinlovinger** [purescript-concur/purescript-concur-react#40]
Say that I have a widget that runs in parallel with an `Effect`, such as
```purs
canvas [ _id $ show canvasId ] []
  <|> (liftAff (runCanvas canvasId) *> empty)
```

I'm finding that, when this widget is removed, the `runCanvas` effect continues to run, and as this widget is removed and re-added, the effects stack, resulting is performance slowdown and memory buildup.

Does Concur have something like React's `componentWillUnmount` to clean up effects,  like the example above.

**Comment by pkamenarsky** [comment]
I find myself in need of "widget finalisers" as well. My particular use case is a drag & drop implementation, for which I need to add event handlers to the `document` directly (so that mouse movements are registered outside the element, as well as browser). The problem is that without finalisers there's no way to remove the previously added event handlers if a `Widget` that's in the middle of a drag & drop cycle gets killed.

Maybe something like:

```
withFinaliser :: forall a. Effect Unit -> Widget v a -> Widget v a
```

where the effect would be called if the `Widget` itself *or any of its parents* gets killed.

**Comment by ajnsit** [comment]
Two developments -

1. I have started working on a new [VDom backend for Concur](https://github.com/purescript-concur/purescript-concur-vdom) which uses [Halogen-VDom](https://github.com/purescript-halogen/purescript-halogen-vdom). With this, I think we'll also be able to have components with internal persistent state, and initialisers and finalisers. I am still coming up with an appropriate API.

2. I will be adding support for internal state, and React lifecycle methods such as componentWillUnmount to the React backend as well.

I hope to have the same unified API for both of these things.

**Comment by ajnsit** [comment]
We are going to track the lifecycle and local state implementation in an umbrella issue - https://github.com/purescript-concur/purescript-concur-react/issues/46. Closing this in favor of that issue.
